File: sync/src/main/java/com/inceptionnotes/sync/ws/WebsocketServer.java
package com.inceptionnotes.sync.ws;

import com.inceptionnotes.sync.Server;
import com.inceptionnotes.sync.events.Event;

import java.util.HashSet;
import java.util.Set;

import javax.websocket.HandshakeResponse;
import javax.websocket.server.HandshakeRequest;
import javax.websocket.server.ServerEndpointConfig;

/**
 * Created by jacob on 1/20/18.
 */

public class WebsocketServer extends ServerEndpointConfig.Configurator {

    private final Server server;
    private final Set<WebsocketClient> sessions = new HashSet<>();

    public WebsocketServer() {
        server = new Server(this);
    }

    public void broadcast(WebsocketClient client, Event event) {
        synchronized (sessions) {
            for (WebsocketClient other : sessions) {
                if (!other.getSession().isOpen()) {
                    continue;
                }

                if (client.getSession().getId().equals(other.getSession().getId())) {
                    continue;
                }

                other.getClient().send(event);
            }
        }
    }

    public void join(WebsocketClient client) {
        synchronized (sessions) {
            sessions.add(client);
        }
    }

    public void leave(WebsocketClient client) {
        synchronized (sessions) {
            sessions.remove(client);
        }
    }

    @Override
    public void modifyHandshake(ServerEndpointConfig conf, HandshakeRequest req, HandshakeResponse resp) {
        conf.getUserProperties().put("server", server);
    }
}

File: sync/src/main/java/com/inceptionnotes/sync/ws/WebsocketServer.kt
package com.inceptionnotes.sync.ws

import com.inceptionnotes.sync.world.Server
import com.queatz.on.On
import javax.websocket.HandshakeResponse
import javax.websocket.server.HandshakeRequest
import javax.websocket.server.ServerEndpointConfig

/**
 * Created by jacob on 1/20/18.
 */

class WebsocketServer : ServerEndpointConfig.Configurator() {

    companion object {
        internal val on = On()
    }

    override fun modifyHandshake(conf: ServerEndpointConfig, req: HandshakeRequest, resp: HandshakeResponse) {
        conf.userProperties["server"] = on<Server>()
    }
}
